bpf: Define new CILIUM_NOTIFY_TRACE perf ring buffer msg type

rlenglet opened the pull request.

rlenglet committed:
bpf: Rename debug tracing functions

Rename cilium_trace* into cilium_dbg* to reflect the actual
message type's name.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
bpf: Define new CILIUM_NOTIFY_TRACE perf ring buffer msg type

Define CILIUM_NOTIFY_TRACE with the same struct as
CILIUM_NOTIFY_DROP, as a replacement for the DBG_CAPTURE_DELIVERY
debug capture type, to collect additional information about
forwarded packets:
* Source security ID.
* Destination security ID.
* Destination endpoint ID.

Replace DBG_CAPTURE_DELIVERY dbg messages with trace messages in
bpf_lxc.c.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
bpf: Use configurable trace capture length for DROP events

Signed-off-by: Romain Lenglet <romain@covalent.io>

tgraf commented on bpf/lib/drop.h:
```diff
@@ -36,7 +36,7 @@
 __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_DROP_NOTIFY) int __send_drop_notify(struct __sk_buff *skb)
 {
-	uint64_t skb_len = skb->len, cap_len = min(64ULL, skb_len);
+	uint64_t skb_len = skb->len, cap_len = min(TRACE_PAYLOAD_LEN, skb_len);
```
👍

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -265,7 +266,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 			return ret;
 
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
`TRACE_TO_PROXY`

rlenglet replied:
Could it be better to include both source and destination in the observation point names, e.g. `TRACE_LXC_TO_PROXY`?

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -368,7 +369,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 
 #ifndef POLICY_ENFORCEMENT
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
`TRACE_TO_HOST`

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -395,7 +396,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 #ifndef POLICY_ENFORCEMENT
 	/* No policy, pass directly down to stack */
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, 0);
+	send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, 0);
```
`TRACE_TO_STACK`

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -587,7 +588,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 			return ret;
 
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
`TRACE_TO_PROXY`

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -677,7 +678,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 
 #ifndef POLICY_ENFORCEMENT
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
`TRACE_TO_HOST`

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -706,7 +707,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 #ifndef POLICY_ENFORCEMENT
 	/* No policy, pass directly down to stack */
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, 0);
+	send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, 0);
```
`TRACE_TO_STACK`

rlenglet replied:
Done.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -1024,9 +1025,7 @@ __section_tail(CILIUM_MAP_POLICY, LXC_ID) int handle_policy(struct __sk_buff *sk
-	ifindex = skb->cb[CB_IFINDEX];
-
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, ifindex);
+	send_trace_notify(skb, TRACE_TO_LXC, src_label, SECLABEL, LXC_ID, ifindex);
```
👍

tgraf commented on daemon/main.go:
```diff
@@ -447,6 +447,7 @@ func initEnv() {
 	}
 
 	config.Opts.Set(endpoint.OptionDropNotify, true)
+	config.Opts.Set(endpoint.OptionTraceNotify, true)
```
👍

aanm commented on cilium/cmd/endpoint_config.go:
```diff
@@ -30,7 +30,7 @@ var listOptions bool
 	Use:     "config <endpoint id> [<option>=(enable|disable) ...]",
 	Short:   "View & modify endpoint configuration",
-	Example: "endpoint config 5421 DropNotification=disable",
+	Example: "endpoint config 5421 DropNotification=false TraceNotification=false",
```
As a user: What's the difference between the Trace and the Debug?

tgraf replied:
Should be simple: Debug should be a developer only flag. We should remove it from the examples. Trace is for end users as well. The overhead of debug will be higher than the trace.

rlenglet replied:
I think the naming is confusing. The `CILIUM_NOTIFY_DBG_CAPTURE` messages are sent if `ENABLE_TRACE` is defined. That is currently always defined, and it's defined independently from `DEBUG`. `ENABLE_TRACE` can be undefined if the `--disable-trace` is given to Cilium. However, that option is used nowhere:
```
$ grep -r disable-trace *
daemon/main.go:		"disable-trace", false, "Disable ability to trace datapath actions")
```
Therefore, we currently always send the `CILIUM_NOTIFY_DBG_CAPTURE` messages, which I believe is a bug.

As @tgraf mentioned, the new TraceNotification is to enable a new set of messages, which are not only for debugging, but are as useful to users as drop messages.

I propose to remove the `ENABLE_TRACE` define, and use `DEBUG` instead to enable the `CILIUM_NOTIFY_DBG_CAPTURE` messages. I'll also remove the currently unused `--disable-trace` option.

rlenglet replied:
Pushed a new commit to replace `ENABLE_TRACE` with `DEBUG` and remove `disable-trace`.

rlenglet committed:
bpf: Use DEBUG instead of ENABLE_TRACE for debug captures

Enable debug capture messages if DEBUG is defined, instead of
ENABLE_TRACE which is always defined.
Remove the unused disable-trace option which allowed undefining
ENABLE_TRACE.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet commented on pkg/bpfdebug/trace.go:
```diff
@@ -0,0 +1,78 @@
+		GetConnectionSummary(data[TraceNotifyLen:]))
+}
+
+// Dump prints the trace notification in human readable form
```
Done.

rlenglet commented on pkg/bpfdebug/trace.go:
```diff
@@ -0,0 +1,78 @@
+		GetConnectionSummary(data[TraceNotifyLen:]))
+}
+
+// Dump prints the trace notification in human readable form
```
Done.

rlenglet committed:
tests: Fix dump_cli_output helper to handle multiple labels

Signed-off-by: Romain Lenglet <romain@covalent.io>
